cggmp21/src/signing.rs

```diff
@@ -987,93 +1114,86 @@ where
-                msg_id,
-                (psi_invalid, hat_psi_invalid, psi_prime_invalid),
-            ))
-        }
-        runtime.yield_now().await;
-    }
+    let faulty_parties =
+        utils::collect_blame_with_data(&round2_msgs, &ciphertexts, |j, msg, ciphertexts| {
+            tracer.stage("Retrieve auxiliary data");
+            let X_j = X[usize::from(j)];
+            let enc_j = &N[usize::from(j)];
+
+            tracer.stage("Validate tilde_psi_j");
+            let tilde_psi_invalid = pi_elog::non_interactive::verify::<E, D>(
+                &unambiguous::ProofElog {
+                    sid,
+                    prover: j,
+                    prime: false,
+                },
+                pi_elog::Data {
+                    l: &ciphertexts.B1,
+                    m: &ciphertexts.B2,
+                    x: &ciphertexts.Y,
+                    y: &msg.Gamma,
+                    h: &Point::generator().to_point(),
+                },
+                &msg.tilde_psi.0,
+                &msg.tilde_psi.1,
+            )
+            .err();
+
+            tracer.stage("Validate psi_i,j");
+            let psi_j_invalid = pi_aff::non_interactive::verify::<E, D>(
+                &unambiguous::ProofPsi {
+                    sid,
+                    prover: j,
+                    hat: false,
+                },
+                &R_i.into(),
+                pi_aff::Data {
+                    key_j: &dec_i,
+                    key_i: enc_j,
+                    c: &K_i,
+                    d: &msg.D,
+                    y: &msg.F,
+                    x: &msg.Gamma,
+                },
+                &msg.psi_j.0,
+                &security_params.pi_aff,
+                &msg.psi_j.1,
+            )
+            .err();
+
+            tracer.stage("Validate hat_psi_i,j");
+            let hat_psi_j_invalid = pi_aff::non_interactive::verify::<E, D>(
+                &unambiguous::ProofPsi {
+                    sid,
+                    prover: j,
+                    hat: true,
+                },
+                &R_i.into(),
+                pi_aff::Data {
+                    key_j: &dec_i,
+                    key_i: enc_j,
+                    c: &K_i,
+                    d: &msg.hat_D,
+                    y: &msg.hat_F,
+                    x: &X_j,
+                },
+                &msg.hat_psi_j.0,
+                &security_params.pi_aff,
+                &msg.hat_psi_j.1,
+            )
+            .err();
 
+            if tilde_psi_invalid.is_some() || psi_j_invalid.is_some() || hat_psi_j_invalid.is_some()
+            {
+                Some((tilde_psi_invalid, psi_j_invalid, hat_psi_j_invalid))
+            } else {
+                None
+            }
+        });
     if !faulty_parties.is_empty() {
         return Err(SigningAborted::InvalidPsi(faulty_parties).into());
     }
+    runtime.yield_now().await;
 
     // Step 2
     tracer.stage("Compute Gamma, Delta_i, delta_i, chi_i");
```

Comment:
right below there is this line:
`let Gamma = Gamma_i + round2_msgs.iter().map(|msg| msg.Gamma).sum::<Point<E>>();`

In the spec there is no such addition. Maybe it's normal because `round2_msgs` is built excluding `i` index?

Reply:
Yes, exactly `round2_msgs` only includes messages received from other parties, so we need to add our Gamma_i